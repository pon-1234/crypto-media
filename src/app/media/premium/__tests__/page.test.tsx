/**
 * @vitest-environment jsdom
 */
import { describe, it, expect, vi } from 'vitest'
import PremiumArticlesPage from '../page'
import {
  getMediaArticlesByMembershipLevel,
  getCategories,
  getTags,
} from '@/lib/microcms'
import type { MediaArticle, Category, Tag } from '@/lib/schema'

vi.mock('@/lib/microcms', () => ({
  getMediaArticlesByMembershipLevel: vi.fn(),
  getCategories: vi.fn(),
  getTags: vi.fn(),
}))

vi.mock('@/components/media/ArticleCard', () => ({
  ArticleCard: vi.fn(({ article }) => (
    <div data-testid={`article-${article.id}`}>{article.title}</div>
  )),
}))

vi.mock('@/components/ui/Pagination', () => ({
  Pagination: vi.fn(({ currentPage, totalPages }) => (
    <div data-testid="pagination">
      Page {currentPage} of {totalPages}
    </div>
  )),
}))

vi.mock('@/components/media/ArticleFilters', () => ({
  ArticleFilters: vi.fn(({ selectedCategory, selectedTag }) => (
    <div data-testid="filters">
      Category: {selectedCategory || 'none'}, Tag: {selectedTag || 'none'}
    </div>
  )),
}))

describe('PremiumArticlesPage', () => {
  const mockArticles: MediaArticle[] = [
    {
      id: '1',
      title: 'プレミアム記事1',
      slug: 'premium-1',
      type: 'article',
      membershipLevel: 'paid',
      content: '有料コンテンツ1',
      heroImage: {
        url: 'https://example.com/image1.jpg',
        width: 1200,
        height: 630,
      },
      createdAt: '2024-01-01T00:00:00Z',
      updatedAt: '2024-01-01T00:00:00Z',
      publishedAt: '2024-01-01T00:00:00Z',
      revisedAt: '2024-01-01T00:00:00Z',
    },
    {
      id: '2',
      title: 'プレミアム記事2',
      slug: 'premium-2',
      type: 'survey_report',
      membershipLevel: 'paid',
      content: '有料コンテンツ2',
      heroImage: {
        url: 'https://example.com/image2.jpg',
        width: 1200,
        height: 630,
      },
      createdAt: '2024-01-02T00:00:00Z',
      updatedAt: '2024-01-02T00:00:00Z',
      publishedAt: '2024-01-02T00:00:00Z',
      revisedAt: '2024-01-02T00:00:00Z',
    },
  ]

  const mockCategories: Category[] = [
    {
      id: 'cat1',
      name: 'DeFi',
      slug: 'defi',
      createdAt: '2024-01-01T00:00:00Z',
      updatedAt: '2024-01-01T00:00:00Z',
    },
  ]

  const mockTags: Tag[] = [
    {
      id: 'tag1',
      name: 'Bitcoin',
      slug: 'bitcoin',
      createdAt: '2024-01-01T00:00:00Z',
      updatedAt: '2024-01-01T00:00:00Z',
    },
  ]

  it('有料記事一覧を正しく表示する', async () => {
    vi.mocked(getMediaArticlesByMembershipLevel).mockResolvedValue({
      contents: mockArticles,
      totalCount: 2,
      offset: 0,
      limit: 12,
    })
    vi.mocked(getCategories).mockResolvedValue({
      contents: mockCategories,
      totalCount: 1,
      offset: 0,
      limit: 100,
    })
    vi.mocked(getTags).mockResolvedValue({
      contents: mockTags,
      totalCount: 1,
      offset: 0,
      limit: 100,
    })

    const page = await PremiumArticlesPage({ searchParams: {} })

    expect(getMediaArticlesByMembershipLevel).toHaveBeenCalledWith('paid', {
      limit: 12,
      offset: 0,
      orders: '-publishedAt',
      filters: 'membershipLevel[equals]paid',
    })
    expect(page).toMatchSnapshot()
  })

  it('ページネーションが正しく動作する', async () => {
    vi.mocked(getMediaArticlesByMembershipLevel).mockResolvedValue({
      contents: mockArticles,
      totalCount: 30,
      offset: 12,
      limit: 12,
    })
    vi.mocked(getCategories).mockResolvedValue({
      contents: mockCategories,
      totalCount: 1,
      offset: 0,
      limit: 100,
    })
    vi.mocked(getTags).mockResolvedValue({
      contents: mockTags,
      totalCount: 1,
      offset: 0,
      limit: 100,
    })

    const page = await PremiumArticlesPage({ searchParams: { page: '2' } })

    expect(getMediaArticlesByMembershipLevel).toHaveBeenCalledWith('paid', {
      limit: 12,
      offset: 12,
      orders: '-publishedAt',
      filters: 'membershipLevel[equals]paid',
    })
    expect(page).toMatchSnapshot()
  })

  it('カテゴリフィルタが適用される', async () => {
    vi.mocked(getCategories)
      .mockResolvedValueOnce({
        contents: [mockCategories[0]],
        totalCount: 1,
        offset: 0,
        limit: 1,
      })
      .mockResolvedValueOnce({
        contents: mockCategories,
        totalCount: 1,
        offset: 0,
        limit: 100,
      })

    vi.mocked(getMediaArticlesByMembershipLevel).mockResolvedValue({
      contents: mockArticles,
      totalCount: 2,
      offset: 0,
      limit: 12,
    })
    vi.mocked(getTags).mockResolvedValue({
      contents: mockTags,
      totalCount: 1,
      offset: 0,
      limit: 100,
    })

    await PremiumArticlesPage({ searchParams: { category: 'defi' } })

    expect(getCategories).toHaveBeenCalledWith({ filters: 'slug[equals]defi' })
    expect(getMediaArticlesByMembershipLevel).toHaveBeenCalledWith('paid', {
      limit: 12,
      offset: 0,
      orders: '-publishedAt',
      filters: 'membershipLevel[equals]paid[and]category[equals]cat1',
    })
  })

  it('指定されたカテゴリが存在しない場合、カテゴリフィルタは適用されない', async () => {
    vi.mocked(getCategories)
      .mockResolvedValueOnce({
        contents: [],
        totalCount: 0,
        offset: 0,
        limit: 1,
      }) // No category found
      .mockResolvedValueOnce({
        contents: mockCategories,
        totalCount: 1,
        offset: 0,
        limit: 100,
      }) // For filters

    vi.mocked(getMediaArticlesByMembershipLevel).mockResolvedValue({
      contents: mockArticles,
      totalCount: 2,
      offset: 0,
      limit: 12,
    })
    vi.mocked(getTags).mockResolvedValue({
      contents: mockTags,
      totalCount: 1,
      offset: 0,
      limit: 100,
    })

    await PremiumArticlesPage({ searchParams: { category: 'unknown' } })

    expect(getMediaArticlesByMembershipLevel).toHaveBeenCalledWith('paid', {
      limit: 12,
      offset: 0,
      orders: '-publishedAt',
      filters: 'membershipLevel[equals]paid',
    })
  })

  it('指定されたタグが存在しない場合、タグフィルタは適用されない', async () => {
    vi.mocked(getTags)
      .mockResolvedValueOnce({
        contents: [],
        totalCount: 0,
        offset: 0,
        limit: 1,
      }) // No tag found
      .mockResolvedValueOnce({
        contents: mockTags,
        totalCount: 1,
        offset: 0,
        limit: 100,
      }) // For filters

    vi.mocked(getMediaArticlesByMembershipLevel).mockResolvedValue({
      contents: mockArticles,
      totalCount: 2,
      offset: 0,
      limit: 12,
    })
    vi.mocked(getCategories).mockResolvedValue({
      contents: mockCategories,
      totalCount: 1,
      offset: 0,
      limit: 100,
    })

    await PremiumArticlesPage({ searchParams: { tag: 'unknown' } })

    expect(getMediaArticlesByMembershipLevel).toHaveBeenCalledWith('paid', {
      limit: 12,
      offset: 0,
      orders: '-publishedAt',
      filters: 'membershipLevel[equals]paid',
    })
  })

  it('記事が0件の場合、メッセージを表示する', async () => {
    vi.mocked(getMediaArticlesByMembershipLevel).mockResolvedValue({
      contents: [],
      totalCount: 0,
      offset: 0,
      limit: 12,
    })
    vi.mocked(getCategories).mockResolvedValue({
      contents: mockCategories,
      totalCount: 1,
      offset: 0,
      limit: 100,
    })
    vi.mocked(getTags).mockResolvedValue({
      contents: mockTags,
      totalCount: 1,
      offset: 0,
      limit: 100,
    })

    const page = await PremiumArticlesPage({ searchParams: {} })

    expect(page).toMatchSnapshot()
  })

  it('エラーが発生した場合、エラーをスローする', async () => {
    const error = new Error('API Error')
    vi.mocked(getMediaArticlesByMembershipLevel).mockRejectedValue(error)

    await expect(PremiumArticlesPage({ searchParams: {} })).rejects.toThrow(
      'API Error'
    )
  })
})
