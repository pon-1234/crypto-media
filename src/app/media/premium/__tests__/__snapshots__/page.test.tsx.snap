// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PremiumArticlesPage > ページネーションが正しく動作する 1`] = `
<div
  className="min-h-screen bg-gray-50"
>
  <div
    className="mx-auto max-w-7xl px-4 py-8 sm:px-6 sm:py-12 lg:px-8"
  >
    <div
      className="mb-8 sm:mb-12"
    >
      <h1
        className="mb-4 text-2xl font-bold text-gray-900 sm:text-3xl md:text-4xl"
      >
        プレミアム記事
      </h1>
      <p
        className="max-w-3xl text-base text-gray-600 sm:text-lg"
      >
        有料会員限定の深掘り記事や詳細な分析レポートをお届けします。
      </p>
    </div>
    <div
      className="mb-8 lg:mb-12"
    >
      <spy
        categories={
          [
            {
              "createdAt": "2024-01-01T00:00:00Z",
              "id": "cat1",
              "name": "DeFi",
              "slug": "defi",
              "updatedAt": "2024-01-01T00:00:00Z",
            },
          ]
        }
        tags={
          [
            {
              "createdAt": "2024-01-01T00:00:00Z",
              "id": "tag1",
              "name": "Bitcoin",
              "slug": "bitcoin",
              "updatedAt": "2024-01-01T00:00:00Z",
            },
          ]
        }
      />
    </div>
    <React.Fragment>
      <div
        className="mb-8 grid grid-cols-1 gap-6 sm:gap-8 md:grid-cols-2 lg:grid-cols-3"
      >
        <spy
          article={
            {
              "content": "有料コンテンツ1",
              "createdAt": "2024-01-01T00:00:00Z",
              "heroImage": {
                "height": 630,
                "url": "https://example.com/image1.jpg",
                "width": 1200,
              },
              "id": "1",
              "membershipLevel": "paid",
              "publishedAt": "2024-01-01T00:00:00Z",
              "revisedAt": "2024-01-01T00:00:00Z",
              "slug": "premium-1",
              "title": "プレミアム記事1",
              "type": "article",
              "updatedAt": "2024-01-01T00:00:00Z",
            }
          }
        />
        <spy
          article={
            {
              "content": "有料コンテンツ2",
              "createdAt": "2024-01-02T00:00:00Z",
              "heroImage": {
                "height": 630,
                "url": "https://example.com/image2.jpg",
                "width": 1200,
              },
              "id": "2",
              "membershipLevel": "paid",
              "publishedAt": "2024-01-02T00:00:00Z",
              "revisedAt": "2024-01-02T00:00:00Z",
              "slug": "premium-2",
              "title": "プレミアム記事2",
              "type": "survey_report",
              "updatedAt": "2024-01-02T00:00:00Z",
            }
          }
        />
      </div>
      <div
        className="mt-8"
      >
        <div
          className="mb-4 text-center"
        >
          <p
            className="text-sm text-gray-600 sm:text-base"
          >
            全
            30
            件中
            13
            -
            14
            件を表示
          </p>
        </div>
        <spy
          className="mt-8"
          currentPage={2}
          totalPages={3}
        />
      </div>
    </React.Fragment>
  </div>
</div>
`;

exports[`PremiumArticlesPage > 有料記事一覧を正しく表示する 1`] = `
<div
  className="min-h-screen bg-gray-50"
>
  <div
    className="mx-auto max-w-7xl px-4 py-8 sm:px-6 sm:py-12 lg:px-8"
  >
    <div
      className="mb-8 sm:mb-12"
    >
      <h1
        className="mb-4 text-2xl font-bold text-gray-900 sm:text-3xl md:text-4xl"
      >
        プレミアム記事
      </h1>
      <p
        className="max-w-3xl text-base text-gray-600 sm:text-lg"
      >
        有料会員限定の深掘り記事や詳細な分析レポートをお届けします。
      </p>
    </div>
    <div
      className="mb-8 lg:mb-12"
    >
      <spy
        categories={
          [
            {
              "createdAt": "2024-01-01T00:00:00Z",
              "id": "cat1",
              "name": "DeFi",
              "slug": "defi",
              "updatedAt": "2024-01-01T00:00:00Z",
            },
          ]
        }
        tags={
          [
            {
              "createdAt": "2024-01-01T00:00:00Z",
              "id": "tag1",
              "name": "Bitcoin",
              "slug": "bitcoin",
              "updatedAt": "2024-01-01T00:00:00Z",
            },
          ]
        }
      />
    </div>
    <React.Fragment>
      <div
        className="mb-8 grid grid-cols-1 gap-6 sm:gap-8 md:grid-cols-2 lg:grid-cols-3"
      >
        <spy
          article={
            {
              "content": "有料コンテンツ1",
              "createdAt": "2024-01-01T00:00:00Z",
              "heroImage": {
                "height": 630,
                "url": "https://example.com/image1.jpg",
                "width": 1200,
              },
              "id": "1",
              "membershipLevel": "paid",
              "publishedAt": "2024-01-01T00:00:00Z",
              "revisedAt": "2024-01-01T00:00:00Z",
              "slug": "premium-1",
              "title": "プレミアム記事1",
              "type": "article",
              "updatedAt": "2024-01-01T00:00:00Z",
            }
          }
        />
        <spy
          article={
            {
              "content": "有料コンテンツ2",
              "createdAt": "2024-01-02T00:00:00Z",
              "heroImage": {
                "height": 630,
                "url": "https://example.com/image2.jpg",
                "width": 1200,
              },
              "id": "2",
              "membershipLevel": "paid",
              "publishedAt": "2024-01-02T00:00:00Z",
              "revisedAt": "2024-01-02T00:00:00Z",
              "slug": "premium-2",
              "title": "プレミアム記事2",
              "type": "survey_report",
              "updatedAt": "2024-01-02T00:00:00Z",
            }
          }
        />
      </div>
    </React.Fragment>
  </div>
</div>
`;

exports[`PremiumArticlesPage > 記事が0件の場合、メッセージを表示する 1`] = `
<div
  className="min-h-screen bg-gray-50"
>
  <div
    className="mx-auto max-w-7xl px-4 py-8 sm:px-6 sm:py-12 lg:px-8"
  >
    <div
      className="mb-8 sm:mb-12"
    >
      <h1
        className="mb-4 text-2xl font-bold text-gray-900 sm:text-3xl md:text-4xl"
      >
        プレミアム記事
      </h1>
      <p
        className="max-w-3xl text-base text-gray-600 sm:text-lg"
      >
        有料会員限定の深掘り記事や詳細な分析レポートをお届けします。
      </p>
    </div>
    <div
      className="mb-8 lg:mb-12"
    >
      <spy
        categories={
          [
            {
              "createdAt": "2024-01-01T00:00:00Z",
              "id": "cat1",
              "name": "DeFi",
              "slug": "defi",
              "updatedAt": "2024-01-01T00:00:00Z",
            },
          ]
        }
        tags={
          [
            {
              "createdAt": "2024-01-01T00:00:00Z",
              "id": "tag1",
              "name": "Bitcoin",
              "slug": "bitcoin",
              "updatedAt": "2024-01-01T00:00:00Z",
            },
          ]
        }
      />
    </div>
    <div
      className="py-16 text-center sm:py-24"
    >
      <p
        className="text-gray-600"
      >
        プレミアム記事がありません。
      </p>
    </div>
  </div>
</div>
`;
